Improve amendment management and default votes

ximinez opened the pull request:
## High Level Overview of Change

This is a follow-up / second attempt to https://github.com/ripple/rippled/pull/3458 at changing default voting amendment behavior. It's more comprehensive, and simplifies the process quite a bit, but it is *not* intended to be the definitive answer to the question of how to manage amendment voting on the network.

It is organized into three commits. Each of the commits should build, but I don't guarantee that they'll pass unit tests. I think it makes more sense to review the changes as a single unit, but they're split up in case there are questions of how something evolved, or if the reviewers absolutely *hate* some subset of the changes.
* The first contains the changes from the older PR (accounting for code drift). Adds yet another list of amendments that should not be voted for by default.
* The second refactors "feature" creation code to centralize the naming, specifying whether it's supported, and providing the default voting behavior in one place.
* The third reorganizes the feature access functions so that supported amendments are always provided with their default votes.

As stated above, the scope of this change is intentionally relatively small. It does not address some of the advanced and controversial questions of amendment voting behavior including, but not limited to:
* Vote proxies
* Abstaining from all voting
* Requiring all amendments to be voted on before starting
* Reducing the voting period
* Adding a "no" voting option. (Current options are "yes" and "abstain".)

### Context of Change

There are two significant problems related to amendment management.
1. All validators will vote for any amendment that is supported (added to the `supportedAmendments` list) by default. This makes it impossible to have multiple versions of the software support a feature without voting for it, without manual intervention from validator operators.
2. It's annoying and error-prone to create a new amendment. The name currently has to be copied in at least three different places.

These changes resolve those issues (and only those issues) by
1. Moving the creation of amendments to one place which specifies the name, whether it's supported, and whether rippled will vote for it by default.
2. Improving instructions and adding runtime validity checks. Developers not familiar with the process should have a much easier time creating and managing the settings for amendments.
3. Allowing an amendment to be supported without being voted for by default by all upgraded validators. This will allow completed features to be released and supported for one or more release cycles before they are voted on by default. Operators will still have full control over voting for or against amendments on their own nodes. If UNL validators wait for more than one release before voting for an amendment (either explicitly or by default), then the "amendment blocked" gating can be significantly reduced.

### Type of Change

- [X ] New feature (non-breaking change which adds functionality)
- [X ] Refactor (non-breaking change that only restructures code)
- [X ] Tests (You added tests for code that already exists, or your new feature included in this PR)

## Before / After

### Before

To create an amendment, the name has to be specified (as a string) in two places:
* In Feature.h, the `featureNames` list
* In Feature.cpp a call to `getRegisteredFeature` to define and initialize a `unit256` variable.

Additionally, a variable need to be declared (at least if you want to use it anywhere else in the code):
* In Feature.h an `extern uint256 const` variable declaration

Later, when a feature is development complete, and ready to support, the name also needs to be added (as a string) to
* In Feature.cpp, the `supportedAmendments()` list

There is no way to add an amendment to `supportedAmendments()` and not vote for it without operator intervention.

### After

To create an amendment, the name has to be specified in one place:
* In Feature.cpp a call to `registeredFeature` to define and initialize a `unit256` variable.

Additionally, the `numFeatures` counter must be incremented. However, if this step is skipped, an assert on startup will catch it and remind the developer (reducing potential errors).

The variable still needs to be declared:
* In Feature.h an `extern uint256 const` variable declaration

Later, when a feature is development complete, and ready to support, the existing call to `registerFeature` only needs to have one of it's parameters modified. Another parameter to `registerFeature` specifies the default voting behavior for supported amendments: yes or abstain.

## Test Plan

The "CryptoConditionsSuite" is listed as supported, and has not yet been enabled on the Mainnet, but is "not ready for prime time". However, it cannot be removed, because it could be enabled at any time, and removing it would cause all future versions to be amendment blocked, and cause other problems. Any current node that has not explicitly vetoed it will vote for it. These votes can be captured and verified on any network (main or test) that hasn't enabled it yet.

The updated registration for this amendment specifies it as `Supported::yes, DefaultVote::abstain`, which means it's still supported, but will not be voted for without explicit operator action. Votes captured from updated nodes on any network where it's not enabled will show that amendment _not_ being voted for.

gregtatcam commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -25,41 +26,180 @@
+        {
+        }
+    };
+    std::vector<Feature> features;
```
Consider using `boost::multi_index::multi_index_container`. This way `features`, `featureToIndex`, and `nameToFeature` can be consolidated into one container. Also `index` can be removed from `Feature`. Something like this:

     struct ByName{};
     struct ByFeature{};
     struct Feature {
          std::string name;
          uint256 feature;
     };
     using namespace boost::multi_index;
 
     multi_index_container<
         Feature,
         indexed_by<
             random_access<>,
             hashed_unique<tag<ByFeature>, member<Feature, uint256, &Feature::feature>>,
             hashed_unique<tag<ByName>, member<Feature, std::string, &Feature::name>>
         >
     > features;
     // get by feature
     const auto& feature_index = features.get<1>();
     auto feature_it = feature_index.find(101);
     // get by name
     const auto& name_index = features.get<2>();
     auto name_it = name_index.find("featureName");
     // get by index
     const auto& sequence = features.get<0>();
     auto f = sequence[0];
     // get index from name
     auto it_to = sequence.iterator_to(*name_it);
     auto index = it_to - sequence.begin();

ximinez replied:
Good idea. I had considered using `boost::intrusive::multiset`, but wasn't going to bother unless this got some traction, then forgot about it. I think the `multi_index_container` works even better in this case. Thanks!

scottschurr replied:
Nice!  I've considered using `multi_index_container` to replace uses of the aged containers.  It would be good to get some experience with that container to see how we like it.

gregtatcam reviewed (commented):
There is a compile warning:
`src/ripple/protocol/impl/Feature.cpp:232:21: warning: unused variable ‘feature’ [-Wunused-variable]
  232 |         auto const& feature = features.emplace_back(name, f);`
and a compile error:
`src/ripple/protocol/impl/Feature.cpp:129:53:   required from here
boost_1_75_0/boost/container_hash/extensions.hpp:305:30: error: no matching function for call to ‘hash_value(const ripple::base_uint<256>&)’
  305 |             return hash_value(val);`

ximinez commented:
> There is a compile warning:
> `src/ripple/protocol/impl/Feature.cpp:232:21: warning: unused variable ‘feature’ [-Wunused-variable] 232 | auto const& feature = features.emplace_back(name, f);`
> and a compile error:
> `src/ripple/protocol/impl/Feature.cpp:129:53: required from here boost_1_75_0/boost/container_hash/extensions.hpp:305:30: error: no matching function for call to ‘hash_value(const ripple::base_uint<256>&)’ 305 | return hash_value(val);`

Fixed. Thanks for catching these. Ironically, I actually ran across the same error in VS, and fixed it there, which is why I only had to move code. I didn't follow up with the other builds.

gregtatcam reviewed (approved):
:+1: Nice improvements!

scottschurr commented:
I'm still in the middle of this review.  However I have some significant changes to `AmendmentTable` I'd like to suggest.  I thought I'd go ahead and put these out there so you can think about them.

The biggest change has to do with the way the Wallet db handles `vetoed`. Firstly, despite what our API says, there are no vetoes.  There are only up-votes and down-votes.  No one gets a veto.

Secondly, vetoed is stored as `1` and not-vetoed as `0`.  This means every time I run across this I have to stand on my head.  Am I voting in favor of this or not?

We can't change what is stored in the database, but we can change how we _talk_ about what is stored in the database.

So I suggest we add an `AmendmentVote` enum to RelationalDBInterface_global.h.  We can replace the integer values that are going in and out of the database with this enum, at least at the interface.  I tried this and I felt like it really helped the readability of the `AmendmentTable` code.  Of course one person's repast is another person's poison.  You may not like it.  But I thought I'd bring it up for discussion.

A smaller point is the `getName` and `getName2` generic lambdas in `AmendmentTableImpl::AmendmentTableImpl`.  I found the pair of generic lambdas really hard to figure out.  I had to do some experimenting with the code to understand the (simple) thing that is going on.  I suggest reducing that to a single non-generic lambda.  That was a significant readability improvement for me.

I tried out both of these changes.  If you'd like to look at the results, here's what I got: https://github.com/scottschurr/rippled/commit/bd282fd947277bcc250ca97bd4a314380b17afc7  I'm not insisting that these are good changes.  But I hope you'll look them over and see what you think.  Thanks.

scottschurr commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -17,127 +17,285 @@
 namespace ripple {
 
-//------------------------------------------------------------------------------
+enum class Supported : bool { no = false, yes };
```
I'm a little concerned about declaring `Supported` at `namespace ripple` scope.  There are a huge number of things in the code base that are supported.  In this case we're only talking about whether an `Amendment` or `Feature` is supported.  Yeah, this declaration is only visible to humans in this file.  But the compiler leaks the name `ripple::Supported` out to the entire application.

Consider renaming this to `AmendmentSupported`, which I'll admit is a little unwieldy.  Another option would be to declare it as a member of the `Feature` class and use the class to limit the scope.  Or maybe you have a yet better idea.

I won't insist, I'm just a bit concerned.

ximinez replied:
Wouldn't it only leak the name to parts of the application in the same TU?
I put it in the `detail` namespace. I can't say I'm entirely happy with the way the variable initialization looks now, but it works.

scottschurr replied:
I'm pretty sure the name leaks into the entire executable through the linker.

What you _could_ do is put the declaration in an unnamed namespace like this:
```
namespace
{
    enum class Supported : bool { no = false, yes };
}
```
According to cppreference, for an unnamed name space...
>  Its members have potential scope from their point of declaration to the end of the translation unit, and have internal linkage.

https://en.cppreference.com/w/cpp/language/namespace

I think an unnamed namespace would provide the smallest possible scope.

ximinez replied:
Done.

scottschurr reviewed (commented):
Very nice changes!  The unit test coverage looks pretty good.  And I like that we're getting the first use of `boost::multi_index` into the code base.

I made a few suggestions for changes along the way.  All of these suggestions are open for discussion.  But I'm going to hold off approving this pull request until we've had a chance to talk through the suggestions.

ximinez commented:
Rebased on to 1.8.0-b4. Added new amendments, but did not make any other code changes.

intelliot reviewed (approved):
LGTM

scottschurr commented:
@ximinez, I left some comments over a month ago.  Do you have time to go through those comments?  Once that's done then maybe we can mark this as passed...

ximinez commented:
@scottschurr Soon:tm:... I haven't forgotten, but I've had other things going on that pushed this to the backburner.

ximinez commented on src/ripple/app/main/Application.cpp:
```diff
@@ -1250,27 +1250,29 @@ ApplicationImp::setup()
-        supportedAmendments.append(saHashes);
+        auto const supportedAmendments = []() {
+            auto const& amendments = detail::supportedAmendments();
+            std::vector<AmendmentTable::FeatureInfo> hashes;
```
Cleaned up the names in this entire block

ximinez commented on src/ripple/app/main/Application.cpp:
```diff
@@ -1250,27 +1250,29 @@ ApplicationImp::setup()
+            }
+            return hashes;
+        }();
+        Section const& downVotedAmendments =
```
Thanks!

ximinez commented on src/ripple/app/misc/impl/AmendmentTable.cpp:
```diff
@@ -323,26 +324,33 @@ AmendmentTableImpl::AmendmentTableImpl(
         return createFeatureVotes(*db);
     }();
 
+    auto getName2 = [](auto const& amendment, auto const& name) {
+        if (name.empty())
+            return featureToName(amendment);
+        return name;
+    };
+    auto getName = [&getName2](auto const& amd) {
+        return getName2(amd.first, amd.second);
+    };
```
Fixed by pulling in your experimental commit.

ximinez commented on src/ripple/app/misc/impl/AmendmentTable.cpp:
```diff
@@ -376,7 +384,7 @@ AmendmentTableImpl::AmendmentTableImpl(
             if (detect_conflict.count(a.first) == 0)
             {
-                persistVote(a.first, a.second, true);  // veto
+                persistVote(a.first, getName(a), true);  // veto
```
Fixed by pulling in your experimental commit.

ximinez commented on src/ripple/app/misc/impl/AmendmentTable.cpp:
```diff
@@ -405,7 +419,7 @@ AmendmentTableImpl::AmendmentTableImpl(
+                if (auto s = get(amend_hash, lock))
                 {
                     JLOG(j_.info()) << "Amendment {" << *amendment_name << ", "
                                     << amend_hash << "} is vetoed.";
```
Fixed, but I want to ask if you think this will cause confusion with the `feature` RPC command, which also uses `vetoed`: https://github.com/ximinez/rippled/blob/internalveto2/src/ripple/rpc/handlers/Feature1.cpp#L68

scottschurr replied:
That's a good question.  My feeling is that "vetoed" is an unfortunate left over from a long time ago when there were only five validators.  A single downvote really was pretty close to a veto.  With 41 validators it takes 9 downvotes to block an amendment.  So "downvoted" is more representative of the current state of things than "vetoed".

But I also think that it's not worth changing the API to clear this up. The costs of an API change are very high and need to be justified.  Users can live with "vetoed" in the API, even though the meaning is not precisely in line with the reality.

Given that, I still think it's good for our internal names to be clear that no one really gets "veto" power.

That's just my opinion.

ximinez replied:
Yeah, that works for me.

ximinez commented on src/ripple/app/misc/impl/AmendmentTable.cpp:
```diff
@@ -416,35 +430,34 @@ AmendmentTableImpl::AmendmentTableImpl(
+                auto s = add(amend_hash, lock);
+
+                JLOG(j_.debug()) << "Amendment {" << *amendment_name << ", "
+                                 << amend_hash << "} is un-vetoed.";
```
Fixed

ximinez commented on src/ripple/app/misc/impl/AmendmentTable.cpp:
```diff
@@ -358,7 +366,7 @@ AmendmentTableImpl::AmendmentTableImpl(
         {  // Otherwise transfer config data into the table
             detect_conflict.insert(a.first);
-            persistVote(a.first, a.second, false);  // un-veto
+            persistVote(a.first, getName(a), false);  // un-veto
```
Fixed by pulling in your experimental commit.

ximinez commented on src/ripple/protocol/Feature.h:
```diff
@@ -32,119 +32,63 @@
-    std::vector<uint256> features;
-    boost::container::flat_map<uint256, std::size_t> featureToIndex;
-    boost::container::flat_map<std::string, uint256> nameToFeature;
+enum class DefaultVote : bool { abstain = false, yes };
```
My understanding is that it's the opposite - there is no such thing as a `no` vote. A validator either votes yes by sending the appropriate message, or it doesn't vote, which I think of as an abstention.

scottschurr replied:
This is a good discussion to have.

I see your point.  You're looking at the action of sending a message or the "lack of action" of not sending a message.  The action votes yes.  The lack of action, which you call abstaining, votes no.

I'm looking from a different angle, a perspective where we look at the consequences of the actions.  The current system of amendment voting mandates that all of the servers on the UNL take a position.  Today on the main net that's 41 servers.  Actively sending a vote says, yes, I support the amendment.  Not sending the vote has the effect of casting a no vote.  Any server on the UNL that does not send a vote has the _consequence_ of casting a no vote.

I see abstaining as a different animal.  If I abstain, then I'm saying that I am not participating in the vote.  I either don't care enough to have an opinion, or I trust those who are actively voting to be better informed than I am.  But in either case when I abstain I'm choosing to not participate in the vote.

If we were to implement something like abstaining it would have the effect of reducing the total number of votes on a particular amendment.  If one validator abstained from voting on an amendment then the number of required yes votes would change.

The currently implemented amendment voting system does not allow a validator to abstain.  Every operator of a validator on the UNL takes a position on every amendment, whether the operator wants to or not.  But I could imagine a future where something like abstaining might be useful.  I'd like to save the word "abstain" in case it becomes useful in the future.  Right now I think `yes` and `no` are good terms to reflect the consequences of the choice.

Does that makes sense?  I'm open to further discussion.

ximinez replied:
I gotta admit, that really does make sense.

ximinez commented on src/ripple/protocol/Feature.h:
```diff
@@ -32,119 +32,63 @@
-    uint256 const&
-    bitsetIndexToFeature(size_t i) const;
-};
+/** Amendments that this server won't vote for by default. */
+std::size_t
+numDownVotedAmendments();
```
Instead of removing the function, I added a test. LMK if you think that's the wrong way to go.

scottschurr replied:
I'm fine with leaving it as long as its tested.

ximinez commented on src/ripple/protocol/Feature.h:
```diff
@@ -32,119 +32,63 @@
-/** Amendments that this server supports, but doesn't enable by default */
-std::vector<std::string> const&
-supportedAmendments();
+/** Amendments that this server will vote for by default. */
```
Done

ximinez commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -17,127 +17,285 @@
+    {
+        // If this assertion fails, and you just added a feature, increase the
+        // numFeatures value in Feature.h
+        assert(features.size() < numFeatures);
```
Yeah, that's a good idea. Fixed.

ximinez commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -17,127 +17,285 @@
+        auto const f = sha512Half(Slice(name.data(), name.size()));
+#if DEBUG
+        {
+            // remove this before publishing
```
No. I've removed it. The original code was using the `sha512_half_hasher` class. I changed it to use the `sha512Half()` function, and included this check to make sure I didn't miss anything.

ximinez commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -160,56 +332,72 @@ bitsetIndexToFeature(size_t i)
 
+// All known amendments must be registered either here or below with the
+// "retired" amendments
 uint256 const
```
It makes for some pretty long lines, so LMK what you think. I'm considering using a macro instead, even if it means renaming a few variables.

scottschurr replied:
The long lines don't bother me for data tables.  But I'd also be fine with a macro that would remove a bunch of the boiler plate.  In either case I think setting up nice columns makes it easier to see what's happening.

ximinez replied:
Done. The lines are still long, but much less so.

ximinez commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -17,127 +17,285 @@
-    features.reserve(numFeatures());
-    featureToIndex.reserve(numFeatures());
-    nameToFeature.reserve(numFeatures());
+    struct ByName
```
Done and cleaned up.

ximinez commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -17,127 +17,285 @@
+
+    // Intermediate types to help with readability
+    using feature_indexing = boost::multi_index::indexed_by<
+        boost::multi_index::random_access<>,
```
Why would that be necessary? It works just fine without.

ximinez replied:
Done

ximinez commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -17,127 +17,285 @@
+    {
+        if (i >= features.size())
+            LogicError("Invalid FeatureBitset index");
+        const auto& sequence = features.get<0>();
```
Done

ximinez commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -17,127 +17,285 @@
+    size_t
+    getIndex(Feature const& feature) const
+    {
+        const auto& sequence = features.get<0>();
```
Done

ximinez commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -17,127 +17,285 @@
+    Feature const*
+    getByFeature(uint256 const& feature) const
+    {
+        const auto& feature_index = features.get<1>();
```
Done

ximinez commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -17,127 +17,285 @@
+    Feature const*
+    getByName(std::string const& name) const
+    {
+        const auto& name_index = features.get<2>();
```
Done

ximinez commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -17,127 +17,285 @@
+    DefaultVote vote)
+{
+    assert(!readOnly);
+    assert(support != Supported::no || vote != DefaultVote::yes);
```
Done. Since I converted it to be a `LogicError`, the explanation is in the `LogicError` message.

ximinez commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -17,127 +17,285 @@
+        return f;
+    }
+    else
+        return i->feature;
```
Since the checks are converted to `LogicErrors` instead of asserts, this fallback is now impossible, and I've moved code around to reflect that.

ximinez commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -160,56 +332,72 @@ bitsetIndexToFeature(size_t i)
+    retiredFix1528           = retireFeature("fix1528");
 
 // clang-format on
 
```
Done. I only just noticed you included a commit in this message, so I implemented things slightly differently. LMK if I missed anything.

ximinez commented on src/ripple/app/misc/AmendmentTable.h:
```diff
@@ -36,6 +37,19 @@ namespace ripple {
 public:
+    struct FeatureInfo
+    {
+        FeatureInfo() = delete;
```
Yeah, I wanted to be explicit.

ximinez commented on src/ripple/app/misc/impl/AmendmentTable.cpp:
```diff
@@ -323,26 +324,33 @@ AmendmentTableImpl::AmendmentTableImpl(
 
+    auto getName2 = [](auto const& amendment, auto const& name) {
+        if (name.empty())
+            return featureToName(amendment);
```
I have convinced myself that it's impossible to not have a `name` available at any of the callers, so I've removed the `getName` lambda(s).

* `supported` is built from the known amendments, and thus will always have a name.
* `parseSection()` requires that the line has a name after the amendment ID.
* It is possible to provide a hex amendment ID for an unknown amendment to the `feature` RPC command which will create an entry in the `FeatureVotes` table without a name, but if and when that amendment becomes known (by a later release for example), the known name will override the value in the table.

ximinez commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -17,127 +17,285 @@
-    if (i == nameToFeature.end())
-        return std::nullopt;
-    return i->second;
+    readOnly = true;
```
Fixed in part 1

ximinez commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -17,127 +17,285 @@
+    Feature const* feature = getByName(name);
+    if (feature)
+        return feature->feature;
+    return std::nullopt;
```
Yes, easily.

ximinez commented on src/test/app/AmendmentTable_test.cpp:
```diff
@@ -311,27 +333,30 @@ class AmendmentTable_test final : public beast::unit_test::suite
             uint256 const supportedID = amendmentId(a);
-            BEAST_EXPECT(
+            BEAST_EXPECTS(
                 table->isEnabled(supportedID) ==
-                (allEnabled.find(supportedID) != allEnabled.end()));
+                    (allEnabled.find(supportedID) != allEnabled.end()),
+                a +
+                    (table->isEnabled(supportedID) ? " enabled "
+                                                   : " disabled ") +
+                    (allEnabled.find(supportedID) != allEnabled.end()
+                         ? " found"
+                         : " not found"));
```
Yeah. This cleaned up nicely.

ximinez commented:
@scottschurr I think that all your questions and issues have now been resolved. CC: @intelliot @gregtatcam

scottschurr reviewed (approved):
Looks great! :+1: Thanks for your patience with my review.

manojsdoshi commented:
"Merged as part of #3948"

intelliot commented:
Note: Released in version 1.8.0
